src/server/ua_server.c

```diff
@@ -173,6 +178,133 @@ UA_Server_forEachChildNodeCall(UA_Server *server, UA_NodeId parentNodeId,
     return res;
 }
 
+/********************/
+/* GDS Transaction  */
+/********************/
+
+UA_GDSTransaction*
+UA_GDSTransaction_new(const UA_NodeId sessionId) {
+    UA_GDSTransaction *transaction = (UA_GDSTransaction *) UA_calloc(1, sizeof(UA_GDSTransaction));
+    if(!transaction)
+        return NULL;
+
+    UA_NodeId_copy(&sessionId, &transaction->sessionId);
+
+    return transaction;
+}
+
+UA_CertificateGroup*
+UA_GDSTransaction_getCertificateGroup(UA_GDSTransaction *transaction,
+                                      const UA_CertificateGroup *certGroup) {
+#ifdef UA_ENABLE_ENCRYPTION
+    if(!transaction || !certGroup)
+        return NULL;
+
+    for(size_t i = 0; i < transaction->certGroupSize; i++) {
+        UA_CertificateGroup *group = &transaction->certGroups[i];
+        if(UA_NodeId_equal(&group->certificateGroupId, &certGroup->certificateGroupId))
+            return group;
+    }
+
+    /* If the certGroup does not exist, create a new one */
+    transaction->certGroups = (UA_CertificateGroup*)UA_realloc(transaction->certGroups, (transaction->certGroupSize + 1) * sizeof(UA_CertificateGroup));
+    if(!transaction->certGroups)
+        return NULL;
+
+    transaction->certGroupSize++;
+
+    memset(&transaction->certGroups[transaction->certGroupSize-1], 0, sizeof(UA_CertificateGroup));
+
+    UA_TrustListDataType trustList;
+    UA_TrustListDataType_init(&trustList);
+    trustList.specifiedLists = UA_TRUSTLISTMASKS_ALL;
+    certGroup->getTrustList((UA_CertificateGroup*)(uintptr_t)certGroup, &trustList);
+
+    UA_CertificateGroup_Memorystore(&transaction->certGroups[transaction->certGroupSize-1],
+        (UA_NodeId*)(uintptr_t)&certGroup->certificateGroupId, &trustList, certGroup->logging, NULL);
```

Comment:
This is not good. The `maxTrustListSize` limit for the memory store is defaulted to 64 KB here which in some cases is too low. It would be better, if it could take the same limit as the targeted certificate group...